--- a/buildutils/bundle.py
+++ b/buildutils/bundle.py
@@ -162,7 +162,7 @@ def stage_platform_hpp(zmqroot):
     else:
         info("attempting ./configure to generate platform.hpp")
         
-        p = Popen('./configure', cwd=zmqroot, shell=True,
+        p = Popen('./configure --host i386-linux-gnu', cwd=zmqroot, shell=True,
             stdout=PIPE, stderr=PIPE,
         )
         o,e = p.communicate()
